Feature standard chop

jaurentz opened the pull request:
This pull request is the first step in unifying the chopping of coefficient series across all aspects of Chebfun.

asgeirbirkis commented on @chebfun/odesol.m:
```diff
@@ -65,7 +65,9 @@
 p = chebfunpref();
 p.techPrefs.eps = max(relTol); % Use the same tolerance for each column.
-
+temp = cheboppref;
```
@jaurentz Instead of creating a cheboppref here, I think you should pass the happinessCheck as a part of the OPT argument to this method -- that way, it could get passed from the chebop solveivp method.

asgeirbirkis commented on tests/chebfun/test_norm.m:
```diff
@@ -214,7 +214,7 @@
 Iexact = 2.4419616835794597e-5;
 err = abs(I - Iexact);
-pass(33) = ( err < 1e2*vscale(f)*epslevel(f) );
+pass(33) = ( err < 1e9*vscale(f)*epslevel(f) );
```
@jaurentz This is a huge difference in tolerances, is something going wrong here?

jaurentz replied:
The problem was that epslevel was not completely turned off everywhere. For unbndfuns the epslevels used to be quite large. I fixed the problem with updating epslevel here and subsequently the test tolerance became way too tight.

asgeirbirkis commented on @chebop/solvebvpNonlinear.m:
```diff
@@ -77,7 +77,7 @@
 
 linpref = pref;
-linpref.errTol = pref.errTol/10;
+linpref.errTol = pref.errTol/200;
```
I'm not a huge fan of this line. We're getting detoriated convergence at the end of the Newton iteration compared with before.

My current thinking is that we should update linpref.errTol during the Newton iteration, so that it reflects the expected norm of the update once we get to the quadratic phase.

I've exported the Interior layer, nonlinear from Chebgui. Currently, on the branch, I get the following output:
```
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     7.97e-01     6.72e-01     2.75e-01     72        72 
 02     4.93e-01     6.31e-01     6.88e-01    180       180 
 03     2.93e-01     8.15e-01      1.0000     349       349 
 04     1.53e-01     2.66e-01      1.0000     494       494 
 05     5.10e-02     3.34e-01      1.0000     582       582 
 06     5.79e-03     1.13e-01      1.0000     586       586 
 07     7.66e-05     1.32e-02      1.0000     576       586 
 08     1.36e-08     1.77e-04      1.0000     442       586 
 09     1.60e-11     1.18e-03      1.0000     113       586 
--------------------------------------------------------------
>> norm(N(u))
ans =
   1.1703e-06
```
On development, the output is currently:
```
Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     7.97e-01     6.72e-01     2.75e-01    128       128 
 02     4.93e-01     6.31e-01     6.88e-01    256       256 
 03     2.93e-01     8.15e-01      1.0000     512       512 
 04     1.53e-01     2.66e-01      1.0000     724       724 
 05     5.10e-02     3.34e-01      1.0000     724       724 
 06     5.79e-03     1.13e-01      1.0000     724       724 
 07     7.66e-05     1.32e-02      1.0000     724       724 
 08     1.36e-08     1.77e-04      1.0000    1024      1024 
 09     7.08e-16     5.21e-08      1.0000    1024      1024 
--------------------------------------------------------------
>> norm(N(u))
ans =
   2.9227e-10
```
So we've evidently lost the quadratic convergence of Newton towards the end of the iteration, and the residual is four orders of magnitude worse!

On the other hand, we're getting much nicer lengths of the solutions during the iteration, and smaller length of the updates towards the end.

jaurentz replied:
I agree that the accuracy here is unsatisfactory, but I do not think that tightening the tolerance on the inner solves is the culprit. If anything I think it should be tighter, but we can only go so far before linsolve crashes.

asgeirbirkis commented on @chebDiscretization/testConvergence.m:
```diff
@@ -1,60 +1,62 @@
+%   Output:  
+%      ISDONE: True if the functions passed in are sufficiently resolved.
+%      EPSLEVEL: Apparent resolution accuracy (relative to VSCALE or the
+%      functions' intrinsic scale).
```
@jaurentz Missing description of vscale and cutoff from the output.

asgeirbirkis commented on @chebop/solvebvp.m:
```diff
@@ -219,6 +219,9 @@
     [u, info] = solvebvpNonlinear(N, rhs, L, u0, residual, pref, displayInfo);
+
+% simplify output
+u = simplify(u,pref.errTol);
```
Aha! This seems to be the culprit to much of my worries, see comment in discussion I'm about to write.

asgeirbirkis commented:
OK, regarding the line in chebop/solvebvp I just commented on, this is the same example as above with that line commented out:
```
dom = [0 1];
N = chebop(@(x,u) .01.*diff(u,2)+u.*diff(u)-u, dom);
rhs = 0;
N.bc = @(x,u) [u(0)+7./6; u(1)-3./2];
options = cheboppref();
options.display = 'iter';
u = solvebvp(N, rhs, options)
fprintf('len(u): %i. norm(N(u)): %4.2e.\n', length(u), norm(N(u)))

Iter.   || du ||   Contraction    stepsize   len(du)   len(u)
--------------------------------------------------------------
 01     7.97e-01     6.72e-01     2.75e-01     72        72 
 02     4.93e-01     6.31e-01     6.88e-01    180       180 
 03     2.93e-01     8.15e-01      1.0000     349       349 
 04     1.53e-01     2.66e-01      1.0000     494       494 
 05     5.10e-02     3.34e-01      1.0000     582       582 
 06     5.79e-03     1.13e-01      1.0000     586       586 
 07     7.66e-05     1.32e-02      1.0000     576       586 
 08     1.36e-08     1.77e-04      1.0000     442       586 
 09     1.60e-11     1.18e-03      1.0000     113       586 
--------------------------------------------------------------
len(u): 586. norm(N(u)): 7.37e-09.
```
This is the coefficient plot:
![plotcoeffsu](https://cloud.githubusercontent.com/assets/3543429/8957364/89bfbc0e-35ed-11e5-9b50-d9a31c34b16e.png)

With the line restored, this is the output from the fprintf statement at the end, and the coeffs plot:
```
len(u): 316. norm(N(u)): 1.17e-06.
```
![plotcoeffsu_errtolsimp](https://cloud.githubusercontent.com/assets/3543429/8957392/be80d3e2-35ed-11e5-97c0-9fee27e27b10.png)

But cutting as specified by errTol is too agressive as is evident from the plotcoeffs plots -- we've thrown away too much. If we change the line at the end of solvebvp to
```
u = simplify(u);
```
This is the output from fprintf and plotcoeffs we get
```
len(u): 395. norm(N(u)): 4.98e-09.
```
![plotcoeffsu_simp](https://cloud.githubusercontent.com/assets/3543429/8957458/238833de-35ee-11e5-8e9e-70f8b9f7f350.png)


I suspect that this change would allow us to restore the tighter tolerances in many of the chebop tests that had to be changed on this branch.

aaustin141 commented on tests/chebfun/test_addBreaksAtRoots.m:
```diff
@@ -70,14 +70,14 @@
 % Blow-up function:
 op = @(x) x.^2.*(1-exp(-x.^2))-2;
-f = chebfun(op, dom, 'exps', [2 2]);
+f = chebfun(op, dom, 'exps', [2 2], 'happinessCheck', @classicCheck);
```
Why does this test revert to using `classicCheck`?

jaurentz commented:
@asgeirbirkis I believe I've addressed your issues, the most important being the accuracy of the nonlinear solves. To remedy this I've adjusted the tolerance for simplify at the end of solvebvp to be two digits tighter than pref.errTol. This is consistent with the inner solve tolerance for Newton iterations and has no effect on linear problems.

asgeirbirkis commented:
@jaurentz I tried restoring the tighter tolerances in tests/chebop, currently running the tests, if they pass, I'll push them and be a happy man!
